File: t/VB/OmrProcessingDemo/UI/ResultsPanel/ResultSearchDialog.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms

Partial Public Class ResultSearchDialog
   Inherits Form

   Private dgv As DataGridView

   Public Sub New(ByVal dgv As DataGridView)
      InitializeComponent()
      Me.dgv = dgv
      txtSearch.Focus()
   End Sub

   Private Sub btnFindNext_Click(ByVal sender As Object, ByVal e As EventArgs)
      Dim currentRow As Integer = Math.Max(0, dgv.CurrentCell.RowIndex)
      Dim currentcol As Integer = Math.Max(0, dgv.CurrentCell.ColumnIndex)
      Dim direction As Integer = If(rdbtnForward.Checked, 1, -1)

      If rdbtnSearchbyRows.Checked Then

         While DoRowSearch(currentRow, currentcol, dgv.ColumnCount) = False

            If (currentRow = 0 AndAlso direction = -1) OrElse (direction = 1 AndAlso currentRow = dgv.RowCount - 1) Then
               MessageBox.Show("No other matches found.")
               Return
            End If

            currentRow += direction
            currentcol = If(direction = 1, -1, dgv.ColumnCount)
         End While
      End If

      If rdbtnSearchbyCols.Checked Then

         While DoColumnSearch(currentRow, currentcol, dgv.RowCount) = False

            If (currentcol = 0 AndAlso direction = -1) OrElse (direction = 1 AndAlso currentcol = dgv.ColumnCount - 1) Then
               MessageBox.Show("No results found")
               Return
            End If

            Dim prompt As String = String.Format("No results found.  Continue searching in {0} {1}?", If(rdbtnForward.Checked, "next", "previous"), If(rdbtnSearchbyRows.Checked, "row", "column"))

            If MessageBox.Show(prompt, "Continue?", MessageBoxButtons.YesNo) = DialogResult.Yes Then
               currentcol += direction
               currentRow = If(direction = 1, -1, dgv.RowCount)
            Else
               Exit While
            End If
         End While
      End If
   End Sub

   Private Function DoRowSearch(ByVal x As Integer, ByVal y As Integer, ByVal [stop] As Integer) As Boolean
      Dim direction As Integer = If(rdbtnForward.Checked, 1, -1)
      y += direction

      While y >= 0 AndAlso y < [stop]
         Dim cell As DataGridViewCell = dgv(y, x)

         If IsMatch(cell.Value) Then
            dgv.CurrentCell = cell
            Return True
         End If

         y += direction
      End While

      Return False
   End Function

   Private Function IsMatch(ByVal value As Object) As Boolean
      If value IsNot Nothing AndAlso TypeOf value Is String Then
         Dim toMatch As String = If(chkMatchCase.Checked, txtSearch.Text, txtSearch.Text.ToLowerInvariant())
         Dim matchValue As String = If(chkMatchCase.Checked, value.ToString(), value.ToString().ToLowerInvariant())
         Return toMatch = matchValue
      End If

      Return False
   End Function

   Private Function DoColumnSearch(ByVal x As Integer, ByVal y As Integer, ByVal [stop] As Integer) As Boolean
      Dim direction As Integer = If(rdbtnForward.Checked, 1, -1)
      x += direction

      While x >= 0 AndAlso x < [stop]
         Dim cell As DataGridViewCell = dgv(y, x)

         If IsMatch(cell.Value) Then
            dgv.CurrentCell = cell
            Return True
         End If

         x += direction
      End While

      Return False
   End Function

   Private Sub rdbtnSearchby_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs)
      rdbtnForward.Text = If(rdbtnSearchbyRows.Checked, "&Forward", "&Down")
      rdbtnBackward.Text = If(rdbtnSearchbyRows.Checked, "&Backward", "&Up")
   End Sub
End Class

File: t/CS/OmrProcessingDemo/UI/ResultsPanel/ResultSearchDialog.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace OmrProcessingDemo.UI.ResultsPanel
{
   public partial class ResultSearchDialog : Form
   {
      DataGridView dgv;

      public ResultSearchDialog(DataGridView dgv)
      {
         InitializeComponent();

         this.dgv = dgv;

         txtSearch.Focus();
      }

      private void btnFindNext_Click(object sender, EventArgs e)
      {
         int currentRow = Math.Max(0, dgv.CurrentCell.RowIndex);
         int currentcol = Math.Max(0, dgv.CurrentCell.ColumnIndex);

         int direction = rdbtnForward.Checked ? 1 : -1;

         if (rdbtnSearchbyRows.Checked)
         {
            while (DoRowSearch(currentRow, currentcol, dgv.ColumnCount) == false)
            {
               if ((currentRow == 0 && direction == -1) || (direction == 1 && currentRow == dgv.RowCount - 1))
               {
                  MessageBox.Show("No other matches found.");
                  return;
               }

               currentRow += direction;
               currentcol = direction == 1 ? -1 : dgv.ColumnCount;
            }
         }

         if (rdbtnSearchbyCols.Checked)
         {
            while (DoColumnSearch(currentRow, currentcol, dgv.RowCount) == false)
            {
               if ((currentcol == 0 && direction == -1) || (direction == 1 && currentcol == dgv.ColumnCount - 1))
               {
                  MessageBox.Show("No results found");
                  return;
               }

               string prompt = string.Format("No results found.  Continue searching in {0} {1}?", rdbtnForward.Checked ? "next" : "previous", rdbtnSearchbyRows.Checked ? "row" : "column");

               if (MessageBox.Show(prompt, "Continue?", MessageBoxButtons.YesNo) == DialogResult.Yes)
               {
                  currentcol += direction;
                  currentRow = direction == 1 ? -1 : dgv.RowCount;
               }
               else
               {
                  break;
               }
            }
         }
      }

      /// <summary>
      /// Search the selected row starting at the selected column for matching text
      /// </summary>
      /// <returns>True if found, false otherwise</returns>
      private bool DoRowSearch(int x, int y, int stop)
      {
         int direction = rdbtnForward.Checked ? 1 : -1;

         y += direction;
         while (y >= 0 && y < stop)
         {
            DataGridViewCell cell = dgv[y, x];

            if (IsMatch(cell.Value))
            {
               dgv.CurrentCell = cell;
               return true;
            }

            y += direction;
         }

         return false;
      }

      private bool IsMatch(object value)
      {
         if (value != null && value is string)
         {
            string toMatch = chkMatchCase.Checked ? txtSearch.Text : txtSearch.Text.ToLowerInvariant();
            string matchValue = chkMatchCase.Checked ? value.ToString() : value.ToString().ToLowerInvariant();

            return toMatch == matchValue;
         }

         return false;
      }

      private bool DoColumnSearch(int x, int y, int stop)
      {
         int direction = rdbtnForward.Checked ? 1 : -1;

         x += direction;
         while (x >= 0 && x < stop)
         {
            DataGridViewCell cell = dgv[y, x];
            if (IsMatch(cell.Value))
            {
               dgv.CurrentCell = cell;
               return true;
            }

            x += direction;
         }

         return false;
      }

      private void rdbtnSearchby_CheckedChanged(object sender, EventArgs e)
      {
         rdbtnForward.Text = rdbtnSearchbyRows.Checked ? "&Forward" : "&Down";
         rdbtnBackward.Text = rdbtnSearchbyRows.Checked ? "&Backward" : "&Up";
      }
   }
}
